src/ast/transform/TypeChecker.cpp

```diff
@@ -74,9 +75,34 @@ class TypeCheckerImpl : public Visitor<void> {
     /** Analyse types, clause by clause */
     void run() {
         const Program& program = tu.getProgram();
-        for (auto* clause : program.getClauses()) {
+        for (auto const* clause : program.getClauses()) {
             visit(*clause, *this);
         }
+
+        for (auto const* decl : program.getFunctorDeclarations()) {
+            if (!typeAnalysis.hasValidTypeInfo(*decl)) {
+                // This could happen if the types mentioned int the functor declaration
+                // are not valid
+                continue;
+            }
+
+            // Only stateful functors can use UDTs
```

Comment:
Ideally, we phase stateless functors out. They were a nasty hack for an important project and they were implemented in a rush. We have the legacy option and it would be create to declare them as legacy long term.

Reply:
Why the name "stateless"?  I was under the impression that "stateful" meant that other versions of the same rule couldn't run in parallel at the same time because the record table/symbol table could be modified inside the functor.  In which case, there is value in stateless functors, no?  

I also don't like the name "stateless" because aren't functors supposed to be stateless (i.e. pure)?  :)  Finally, I don't like the name "functor" because I think of categories when I hear it :)  I assume that's terminology borrowed from C++'s "function object", right?  (But that's just bike-shedding, and as they say, there are only 2 hard problems in computer science:  Naming things, distributed systems and off-by-one errors.)

Before making the design "complete", I would like to see some of the following improvements:

1. Could we pass the converted values?  E.g. symbols -> std::strings, and do the same with the return values?
2. Is there are cleaner API for records/ADTs?  This would be easy for the generated code (we *could* make the types!).  We could even generate a header, similar to what say Java does with JNI.  Right now, I know of a one *comment* that explains how ADTs are structured and that's inside of the Writer class.  Probably not ideal for newcomers :).  It's also makes it harder to change the representation in the future, since we're not going through an API but rather forcing everyone to inspect the ADTs manually, hence rely on their current representation.  It's also somewhat tricky => error-prone.
3.  At the very least, pass the values to functors by reference rather than by pointer.  This may require a bit of coaxing to make FFI happy.